File: 프로그래머스/2/42890. 후보키/후보키.py
from itertools import combinations

def solution(relation):
    row_size = len(relation) # 행 수
    col_size = len(relation[0]) # 열 수
    
    def check_unique(cand, relation): 
        temp = [tuple(item[key] for key in cand) for item in relation]
        if len(set(temp)) == len(temp):
            return True
        return False

    def check_min(cand, cand_keys):
        for key in cand_keys:
            if key.issubset(cand):
                return False
        return True
    
    candidates = [] # 후보키 조합
    for size in range(1, col_size + 1):
        for comb in combinations(range(col_size), size):
            if check_unique(comb, relation) and check_min(comb, candidates):
                candidates.append(set(comb))
                
    return len(candidates)

File: 프로그래머스/2/42890. 후보키/후보키.java
import java.util.*;

class Solution {
    public int solution(String[][] relation) {
        int rowSize = relation.length;
        int colSize = relation[0].length;
        
        List<Set<Integer>> candidates = new ArrayList<>();
        
        for (int size = 1; size <= colSize; size++) {
            for (Set<Integer> comb : generateCombinations(colSize, size)) {
                if (checkUnique(comb, relation) && checkMin(comb, candidates)) {
                    candidates.add(comb);
                }
            }
        }
        
        return candidates.size();
    }
    
    private boolean checkUnique(Set<Integer> cand, String[][] relation) {
        Set<List<String>> temp = new HashSet<>();
        for (String[] item : relation) {
            List<String> tuple = new ArrayList<>();
            for (int key : cand) {
                tuple.add(item[key]);
            }
            if (!temp.add(tuple)) {
                return false;
            }
        }
        return true;
    }
    
    private boolean checkMin(Set<Integer> cand, List<Set<Integer>> candKeys) {
        for (Set<Integer> key : candKeys) {
            if (cand.containsAll(key)) {
                return false;
            }
        }
        return true;
    }
    
    private List<Set<Integer>> generateCombinations(int n, int r) {
        List<Set<Integer>> combinations = new ArrayList<>();
        generateCombinationsHelper(new HashSet<>(), 0, n, r, combinations);
        return combinations;
    }
    
    private void generateCombinationsHelper(Set<Integer> current, int start, int n, int r, List<Set<Integer>> result) {
        if (current.size() == r) {
            result.add(new HashSet<>(current));
            return;
        }
        
        for (int i = start; i < n; i++) {
            current.add(i);
            generateCombinationsHelper(current, i + 1, n, r, result);
            current.remove(i);
        }
    }
}